bridges/snowbridge/parachain/pallets/inbound-queue/src/lib.rs

```diff
@@ -0,0 +1,342 @@
+					Error::<T>::Send(SendError::DestinationUnsupported),
+				XcmpSendError::ExceedsMaxMessageSize =>
+					Error::<T>::Send(SendError::ExceedsMaxMessageSize),
+				XcmpSendError::MissingArgument => Error::<T>::Send(SendError::MissingArgument),
+				XcmpSendError::Fees => Error::<T>::Send(SendError::Fees),
+			}
+		}
+	}
+
+	/// The current nonce for each channel
+	#[pallet::storage]
+	pub type Nonce<T: Config> = StorageMap<_, Twox64Concat, ChannelId, u64, ValueQuery>;
+
+	/// The current operating mode of the pallet.
+	#[pallet::storage]
+	#[pallet::getter(fn operating_mode)]
+	pub type OperatingMode<T: Config> = StorageValue<_, BasicOperatingMode, ValueQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Submit an inbound message originating from the Gateway contract on Ethereum
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::submit())]
+		pub fn submit(origin: OriginFor<T>, message: Message) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			ensure!(!Self::operating_mode().is_halted(), Error::<T>::Halted);
+
+			// submit message to verifier for verification
+			T::Verifier::verify(&message.event_log, &message.proof)
+				.map_err(|e| Error::<T>::Verification(e))?;
+
+			// Decode event log into an Envelope
+			let envelope =
+				Envelope::try_from(&message.event_log).map_err(|_| Error::<T>::InvalidEnvelope)?;
+
+			// Verify that the message was submitted from the known Gateway contract
+			ensure!(T::GatewayAddress::get() == envelope.gateway, Error::<T>::InvalidGateway);
+
+			// Retrieve the registered channel for this message
+			let channel =
+				T::ChannelLookup::lookup(envelope.channel_id).ok_or(Error::<T>::InvalidChannel)?;
+
+			// Verify message nonce
+			<Nonce<T>>::try_mutate(envelope.channel_id, |nonce| -> DispatchResult {
+				if *nonce == u64::MAX {
+					return Err(Error::<T>::MaxNonceReached.into())
+				}
+				if envelope.nonce != nonce.saturating_add(1) {
+					Err(Error::<T>::InvalidNonce.into())
+				} else {
+					*nonce = nonce.saturating_add(1);
+					Ok(())
+				}
+			})?;
+
+			// Reward relayer from the sovereign account of the destination parachain
+			// Expected to fail if sovereign account has no funds
+			let sovereign_account = sibling_sovereign_account::<T>(channel.para_id);
+			let delivery_cost = Self::calculate_delivery_cost(message.encode().len() as u32);
+			T::Token::transfer(&sovereign_account, &who, delivery_cost, Preservation::Preserve)?;
+
+			// Decode message into XCM
+			let (xcm, fee) =
+				match inbound::VersionedMessage::decode_all(&mut envelope.payload.as_ref()) {
+					Ok(message) => Self::do_convert(envelope.message_id, message)?,
+					Err(_) => return Err(Error::<T>::InvalidPayload.into()),
+				};
+
+			// We embed fees for xcm execution inside the xcm program using teleports
+			// so we must burn the amount of the fee embedded into the XCM script.
+			T::Token::burn_from(&sovereign_account, fee, Precision::Exact, Fortitude::Polite)?;
```

Comment:
What about taking result of `burn_from` and add it to the `Event::MessageReceived`, I think it could be useful later for indexers or what ever like the "real fee"...
```
let burned = T::Token::burn_from(&sovereign_account, fee, Precision::Exact, Fortitude::Polite)?;

...

Self::deposit_event(Event::MessageReceived {
        channel_id: envelope.channel_id,
        nonce: envelope.nonce,
        message_id,
        fee_burned: burned
});
```